google/cloud/opentelemetry/resource_detector.h

```diff
@@ -0,0 +1,49 @@
+// Copyright 2023 Google LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      https://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#ifndef GOOGLE_CLOUD_CPP_GOOGLE_CLOUD_OPENTELEMETRY_RESOURCE_DETECTOR_H
+#define GOOGLE_CLOUD_CPP_GOOGLE_CLOUD_OPENTELEMETRY_RESOURCE_DETECTOR_H
+
+#include "google/cloud/internal/rest_client.h"
+#include "google/cloud/options.h"
+#include "google/cloud/version.h"
+#include <opentelemetry/sdk/resource/resource_detector.h>
+#include <memory>
+
+namespace google {
+namespace cloud {
+namespace otel {
+GOOGLE_CLOUD_CPP_INLINE_NAMESPACE_BEGIN
+
+/// Make an OpenTelemetry Resource Detector for Google Cloud Platform.
+std::unique_ptr<opentelemetry::sdk::resource::ResourceDetector>
+MakeResourceDetector(Options options = {});
+
+GOOGLE_CLOUD_CPP_INLINE_NAMESPACE_END
+}  // namespace otel
+namespace otel_internal {
+GOOGLE_CLOUD_CPP_INLINE_NAMESPACE_BEGIN
+
+using HttpClientFactory =
```

Comment:
nit-ish-maybe-it-matters: opening the `*_internal` namespaces in a public header is a smell.  This probably belongs in `google/cloud/opentelemetry/internal/make_resource_detector.h`